include/aos/common/tools/string.hpp

```diff
@@ -218,6 +218,49 @@ class String : public Array<char> {
      */
     RetWithError<int64_t> ToInt64() { return static_cast<int64_t>(strtoll(CStr(), nullptr, 10)); }
 
+    /**
+     * Converts hex string to byte array.
+     *
+     * @return RetWithError<SerialNum>.
+     */
+    inline Error ToByteArr(Array<uint8_t>& arr)
+    {
+        if (arr.MaxSize() * 2 < Size()) {
+            return Error::Enum::eNoMemory;
+        }
+
+        static const auto parseHex = [](char ch) {
+            if (ch >= '0' && ch <= '9') {
+                return static_cast<uint8_t>(ch - '0');
+            }
+            if (ch >= 'A' && ch <= 'F') {
+                return static_cast<uint8_t>(ch - 'A' + 10);
+            }
+            return static_cast<uint8_t>(255);
+        };
+
+        for (size_t i = 0; i <= Size(); i += 2) {
+            uint8_t low = parseHex(Get()[i]);
+            if (low > 0xFU) {
+                return AOS_ERROR_WRAP(Error::Enum::eInvalidArgument);
```

Comment:
We don't wrap error in lib part: better to know function that called conversion which leads to error.